### ISSUE TITLE
Mapping accelerometer/gyroscope motion to key or macro

### ISSUE THREAD

**Issue by Flybel** [Ryochan7/DS4Windows#1398]
DS4Windows already has a menu to use the gyroscope for controls like mouse or controller axes. I am looking to map sixaxis movement, particularly the y axis of the accelerometer, to keyboard button presses, so when the value surpasses a threshold, a button is pressed once. In simpler terms: I want to press a button by shaking my controller.

I read in [#423](https://github.com/Ryochan7/DS4Windows/issues/423) that mapping them isn't possible due to limitations of what an Xbox controller can do, but since this does not apply to single button presses, I was wondering if this was ever looked into or is something that will be considered?

**Comment by mika-n** [comment]
You can map gyro events to keyboard keys or macro events. And using gyro deadzone/outputScale/sensitivity tweaks you can control how "sensitive" it is as a digital key event. Or am I misunderstanding something you are trying to do?  
Or did you mean that the kbd key is fired only once and not repeated while the controller is tilted beyond the specified "gyro sensitivity point"?

![DS4Win_GyroToKbd](https://user-images.githubusercontent.com/35538525/87849905-e94e0000-c8f4-11ea-8605-e416eb9a211a.png)

**Comment by Flybel** [comment]
Firstly, yes, I'm trying to fire the key only once when the reading surpasses a threshold. However, in this case I want to use translational movement, not rotational. The idea is not to use it as an axis of control but using a gesture as an additional button.
That menu is what I was referring to when I mentioned gyroscope for controls. These tilting controls aren't really suitable for what I'm trying to do though, which is a shaking movement.

In my example: When holding the controller in a neutral position and shaking it up and down, only one of six readings spikes, which is the accelerometer y axis. In the perfectly upright position, the gyro and other accelerometer readings are still and if the position is not upright, the controller can still be tilted when the movement is stopped. So the existing possibilities aren't a reliable option for this type of movement.
![image](https://i.krz.link/foR7E.gif)

I would need one positive y accelerometer spike to fire a key once. Since a rapid stopping movement will almost always follow the rapid starting movement, there would need to be some form of cooldown. Also I acknowledge the issue of gravitational acceleration. If that can be handled, I think this may be a worthwhile method.

The opportunities of having this possibility are:
(a) having multiple additional inputs using one method, similar to the touchpad which offers 4+4 different primary inputs at the same time (in theory accelerometer can provide up to 6 discrete inputs in form of "positive first" and "negative first" spikes on 3 axes, the same could be done with gyroscope readings for another 6 discrete inputs)
(b) the input method doesn't rely on presses, so it can be used simultaneously with existing inputs (no having to let go of other buttons, triggers or sticks)
(c) the added inputs are very quickly accessible as opposed to secondary bindings which require pressing two buttons

As opposed to this raw approach, a more complex method could be defining movement gestures which don't have to be limited to acceleration/rotation in a single direction. Shaking up and down is the most straightforward gesture I can think of.

**Comment by Ryochan7** [comment]
Not sure there is a need to add some form of gesture support for the Gyro. I never really liked shake to action style mapping when it was used in Wii games.
